### PR TITLE
Remove deprecated codepath for old-style autograd.Function (#30696)

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33956 Remove deprecated codepath for old-style autograd.Function (#30696)**

Differential Revision: [D20167359](https://our.internmc.facebook.com/intern/diff/D20167359)

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -609,69 +609,6 @@ PyObject* process_outputs(PyObject *op_obj, const std::shared_ptr<PyNode>& cdata
 }
 
 // Legacy codepath
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: I think the comment should be removed as well.

File: test/test_autograd.py [hunk 2]
```diff
@@ -1616,31 +1597,7 @@ def __del__(self):
...
-        with warnings.catch_warnings(record=True) as w:
-            z = f(x2)
-        self.assertIn('extending-torch-autograd', str(w[1].message))
-        # I don't really care about the functional correctness of this
-        # part of the test: if you make a change that causes this test
-        # to fail, it's probably OK to just fix this test case to follow
-        # it.  I'm mostly making sure we don't segfault here.
-        y.backward()
-        self.assertEqual(x2.grad, x2)
+            CollectOnDelete().forward(torch.randn(1, requires_grad=True)).backward()
```
Comment (albanD) [inline test/test_autograd.py]: This will raise on error on construction no?

File: torch/autograd/function.py [hunk 3]
```diff
@@ -139,8 +139,11 @@ class Function(with_metaclass(FunctionMeta, _C._FunctionBase, _ContextMethodMixi
         >>>         return grad_output * result
     """
 
-    # only for backward compatibility
```
Comment (albanD) [inline torch/autograd/function.py]: Maybe update the comment above by removing "Each function object is meant to be used only once (in the forward pass)." as it is not relevant.
And add to the code sample:
```
# Use it by calling the apply method:
output = Exp.apply(input)
```

**Review by albanD — COMMENTED** [review]
I am surprised there isn't more tests to fix, let's wait for CI.

File: test/test_autograd.py (hunk 2 above)
Comment (glaringlee) [reply test/test_autograd.py]: no, it is not get called.

Comment (albanD) [reply test/test_autograd.py]: What is this test doing if this is not called?

Comment (glaringlee) [reply test/test_autograd.py]: No, I mean, __call__ is not triggered in CollectOnDelete().

Comment (albanD) [reply test/test_autograd.py]: Is there any reason why we allow instantiating old-style Function?
You might want to error out in `FunctionMeta`'s `__init__` in function.py as well if you detect an old style Function.

Comment (glaringlee) [reply test/test_autograd.py]: Nope, no spicial reason. I am just not sure whether we should disallow people creating Function obj. People can call forward using CollectOnDelete.forward() or CollectOnDelete.apply() directly, but there is nothing wrong for people doing CollectOnDelete().forward() as well. Please advise.

**Review by albanD — APPROVED** [review]
After an offline discussion.
More things could be removed but they are postponed to a following PR.

**Pull request closed without merging**